server/etcdserver/corrupt.go

```diff
@@ -241,6 +252,83 @@ func (cm *corruptionMonitor) periodicCheck() error {
 	return nil
 }
 
+func (cm *corruptionChecker) CompactHashCheck() {
+	cm.lg.Info("starting compact hash check",
+		zap.String("local-member-id", cm.hasher.MemberId().String()),
+		zap.Duration("timeout", cm.hasher.ReqTimeout()),
+	)
+	hashes := cm.uncheckedRevisions()
+	// Assume that revisions are ordered from largest to smallest
+	for i, hash := range hashes {
+		peers := cm.hasher.PeerHashByRev(hash.Revision)
+		if len(peers) == 0 {
+			continue
+		}
+		peersChecked := 0
+		for _, p := range peers {
+			if p.resp == nil || p.resp.CompactRevision != hash.CompactRevision {
```

Comment:
Shouldn't `p.resp.CompactRevision != hash.CompactRevision` be an issue, because compaction operation goes through raft?

Reply:
Request to start compaction goes through raft, however compaction itself is done asynchronously in background.

Reply:
Or better to say that compaction window goes through raft, it's up to member to clean up compacted keys in background.

Reply:
I am not convinced.  Each member is indeed performing the compaction in background or asynchronously, but it doesn't matter.  You get the peers' hash using a parameter `hash.Revision` (line 263), so the `resp.CompactRevision` must be equal to `hash.CompactRevision`, or the resp is nil. 

When a member receives the `getPeerHashKVs` request, it just gets the cached value (`hashStorage.hashes`), which is only updated when a compaction operation finishes. The point is each member's cache data must be exactly the same or a subset of other member values. The case of not matching (KeyValueHash.Revision is the same, but CompactRevision are different) will never happen.

Reply:
My point is `p.resp.CompactRevision != hash.CompactRevision` will never be `true`, no matter it's real or pathological case. If somehow it's `true`, them something must be wrong and most likely it's a bug.

Reply:
Re-reading your code change, it seems that it's possible for `p.resp.CompactRevision != hash.CompactRevision` to be `true`. 

When there is no any compaction historical data, then the member may calculate the hash itself. In this case, the `CompactRevision` might be different. 
[hash.go#L133](https://github.com/serathius/etcd/blob/0dc860e1cf12d4a89953c82cff14e20f50237ae6/server/storage/mvcc/hash.go#L133)

Reply:
Agreed, I implemented it this way to ensure backward compatibility. Just remembered that it's needed to handle upgrade. During upgrade older members will return different CompactRevision.

Does it make sense?

Reply:
OK. The only concern is sometimes it wastes the follower(s)'s CPU to calculate the hash, which isn't used by the leader at all.  But it should be very low possibility to run into this situation.